docs/settings.md
# Logrole Settings

There are two main ways to deploy Logrole.

1. Write all settings to a `config.yml` file (a sample is in
config.sample.yml), then run `logrole_server --config=config.yml`.

2. Set all configuration as environment variables. Run
`logrole_write_config_from_env > config.yml` to write all those environment
variables to a config.yml file. Follow the steps in (1).

### Environment variables

Some environments like Heroku only allow you to set production
configuration via environment variables. Logrole has a second binary,
`logrole_write_config_from_env`, that takes named environment variables and
writes them to a YAML file. You can then run the `logrole_server` binary as
described above.

```
logrole_write_config_from_env > myconfig.yml
logrole_server --config=myconfig.yml
```

To see which environment variables are written to the file, run
`logrole_write_config_from_env --help`. Here is an example:

```
Supported environment variables are:

PORT                   Port to listen on
PUBLIC_HOST            Host your users will browse to to see the site

TWILIO_ACCOUNT_SID     Account SID for your Twilio account
TWILIO_AUTH_TOKEN      Auth token

REALM                  Realm (either "local" or "prod")
TZ                     Default timezone (example "America/Los_Angeles")
TIMEZONES              Comma-separated list of timezones users can choose from
                       (example "America/New_York,UTC,America/Chicago").
                       Defaults to the four US timezones.
EMAIL_ADDRESS          For "Contact Support" on server error pages
PAGE_SIZE              How many resources to fetch/display on each page

SECRET_KEY             64 byte hex key - generate with "openssl rand -hex 32"
MAX_RESOURCE_AGE       How long resources should be visible for - "720h" to
                       hide anything older than 30 days
SHOW_MEDIA_BY_DEFAULT  "false" to hide images behind a toggle when a user
                       browses to a MMS message.

AUTH_SCHEME            "basic", "noop", or "google"
BASIC_AUTH_USER        For basic auth, the username
BASIC_AUTH_PASSWORD    For basic auth, the password
GOOGLE_CLIENT_ID       For Google OAuth
GOOGLE_CLIENT_SECRET   For Google OAuth
GOOGLE_ALLOWED_DOMAINS Comma separated list of domains to allow to
                       authenticate. If empty or omitted, all domains allowed.

ERROR_REPORTER         "sentry", empty, or register your own.
ERROR_REPORTER_TOKEN   Token for the error reporter.

POLICY_FILE            Load policy info from a file
POLICY_URL             Download policy info from the specified URL. HTTPS only.
                       Can be protected with Basic Auth. Consider using Dropbox
                       or Github Gist "raw" URLs.
```

#### Heroku deployment

Logrole comes with a default Procfile and a start script (`bin/serve`) for easy
Heroku deployment. Sensitive environment variables (auth token, basic auth
password, etc) are dropped before the server process starts.

## Settings details

### Secret key

The secret key is used to obscure URL's with sensitive information (such as
NextPageURI's and MMS URL's), to encrypt cookies, and to ensure valid OAuth
sessions. It should be 32 bytes of cryptographically random data. Because some
bytes are unprintable garbage, it's easier to store this value in a file as a
64-byte hex-encoded value.

OpenSSL can generate random bytes for you. Type `openssl rand -hex 32` and you
will get a value like this:

```
73cfe0f6926d3b3600b420dontuse20dbe775c1a8e221c72070e5362516c0a34
```

That's the format of a secret key that you should pass as a `secret_key` to
Logrole.

### Timezones

When was a call or SMS made? You can choose both a default timezone for your
server, and then also a list of timezones that users can select from. All times
will be displayed in that local zone.

Timezones are specified in the format used by [the IANA timezone
database][iana], for example "America/Los_Angeles". [A partial list may be
found at Wikipedia][tz-list]. Note, your server may not have every timezone
listed there.

In a YML file, specify `default_timezone: America/Los_Angeles` to configure
a default timezone, and a list of `timezones` to configure the available
timezones in the menu bar.

```yml
timezones:
  - Asia/Singapore
  - Asia/Tokyo
  - Europe/London
  - America/New_York
  - Africa/Cairo
```

[iana]: https://en.wikipedia.org/wiki/Tz_database
[tz-list]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

## Max Resource Age

You may want to prohibit viewers from seeing a resource older than a certain
age. Specify a `max_resource_age` in your config file to limit the ability to
view resources.

We parse the max resource age using [Go's `time.Duration`
type][parse-duration]. Note that this type does not support days or months,
since a day may have 23 or 25 hours. To specify "30 days", specify the max
resource age as "720h", or `30*24` hours.

```
max_resource_age: 720h
```

[parse-duration]: https://golang.org/pkg/time/#ParseDuration

## Authentication

Logrole supports three different methods of authentication, via the
`auth_scheme` parameter in your YAML file.

### No Authentication

Set `auth_scheme: noop` to let everyone visit your site. By default all
visitors will be able to see everything. (If you are manually configuring
Logrole, you can set the `User` field on the `NoopAuthenticator` to a different
permission set.)

### Basic Authentication

Set `auth_scheme: basic` to enforce access to your site with HTTP Basic Auth.
You can set a basic auth user and password in your config file like so:

```yml
basic_auth_user: test
basic_auth_password: hymanrickover
```

You can specify permissions for the Basic Auth user by adding a policy
file, described below. If no policy is present, permissions for the
[DefaultUser][default-user] are given to all users.

### Google Authentication

Set `auth_scheme: google` to use Google OAuth Authentication. Users will be
redirected to Google to login, and then sent back to Logrole.

You'll need a Google Client ID and Client Secret for OAuth. [Follow these
instructions][google] to get those values.

[google]: https://github.com/saintpete/logrole/tree/master/docs/google.md

#### Allowed domains

You can configure an array of domains that are allowed to access the site.
For example, if you specify "example.com", only email addresses that end with
@example.com will be allowed to access the site.

```yml
google_allowed_domains:
  - example.com
  - example.net
  - example.org
```

## Custom permissions for different groups

Use a `policy` to define groups with different permissions. Your `policy` will
look something like this, in YAML:

```yml
policy:
    - name: support
      default: true
      permissions:
          can_view_message_body: false
          can_play_recordings: false
          can_view_message_price: false
          can_view_call_price: false
      users:
          - test@example.com
          - test@example.net

    - name: engineering
      permissions:
          can_view_call_price: false
      users:
          - eng@example.com
          - eng@example.net
```

Let's walk through that:

- **name:** the group name. Required

- **default:** If a user's email is in the group of allowed_domains, but they're
not explicitly specified in a group in the policy, they'll get the permissions
of a `default` group. Only one group can be the default. If no groups are
marked as default, unknown users will not be allowed to authenticate.

- **permissions:** A list of permissions that this group has. Permissions are
**set to true by default,** so you only need to specify the permissions you
want to disallow. A full list of permissions and descrptions can be found on
[the UserSettings object][user-settings].

- **users:** A list of users in this group. These should match the id provided
  for Basic Auth, or the email address used to sign in with Google. A user
  cannot belong to two different groups.

#### Edge cases

There are two tools for locking down access to your site - configuring the
email domains that are allowed to access the site, and specifying a policy. You
don't have to specify those, and they can interact in surprising ways. We try
our best to do the intuitive thing. Here's a short walkthrough of how Logrole
handles different cases.

If a the user's email address is found in the policy, that user is used.

If no policy is defined, we use google_allowed_domains to determine access, and
return [config.DefaultUser][default-user] for user access for all authenticated
users.

If no policy is defined and google_allowed_domains is empty, we permit all
users to access the site.

If google_allowed_domains is not empty and a user's domain is not allowed to
access the site, they are denied access.

If google_allowed_domains is not empty, and a user's domain is allowed, but
they are not in a group, we use the permissions for the default group. If no
default group exists, the user is denied access.

[user-settings]: https://godoc.org/github.com/saintpete/logrole/config#UserSettings
[default-user]: https://godoc.org/github.com/saintpete/logrole/config#DefaultUser

### What happens to the YAML file?

You don't need to read this if you are just running logrole_server. But if you
want to configure Logrole from another Go app, these details are helpful.

In the `logrole_server` binary three things happen:

1. Parse the YAML file into a config.FileConfig struct in memory.

- From there, call config.NewSettingsFromConfig, which initializes things like
the Twilio Client, and an Authenticator, and gives you back a Settings object.

- From there, call `s, _ := server.NewServer(settings)` and you'll get back
  a HTTP handler. You can then use this to listen on a socket and serve HTTP.

```go
settings := &server.Settings{PublicHost: "example.com", PageSize: 50}
s, err := server.NewServer(settings)
http.Handle("/", s)
http.ListenAndServe(":4114", nil)
```

The logrole_server binary handles these automatically for you. If you are
writing custom code, you can skip any or all of these steps with your own
initialization.

[settings-godoc]: https://godoc.org/github.com/saintpete/logrole/server/#Settings


CHANGELOG.md
# Changes

## 0.72

Reject invalid query parameters on list views

Implement per-user/group MaxResourceAge settings.

Refactor template generation a little bit.

## 0.71

Implement caching for Messages

Show whether a result was returned from the cache, and if so how old that
result is. Alters services.Duration to show fewer bits after the decimal if the
Duration is larger than one second.

## 0.70

Implement search filters for Alerts

## 0.69

Add back caching for calls/messages/conferences

## 0.67

Switch the date filtering system from day-and-UTC based to timezone aware, to
the hour filtering of calls, messages, and conferences.

## 0.66

Cosmetic changes - changing "redacted" to "hidden" and reordering some filters.

## 0.61

Implement an Alerts list view.

## 0.58

Implement multi-user permissions

Specify `policy` or `policy_file` to define permission groups, and define users
to exist in those groups, as well as a default group for unknown users. Support
policy in the GoogleAuthenticator and the BasicAuthAuthenticator. Document how
policy behaves if it is/isn't specified, and how it interacts with
allowed_domains.

Moves all of the Authenticator code into the config directory. The interaction
between GoogleAuthenticator and the server directory is a little complicated -
we want to render a 401 error if Google auth fails, which needs to be done from
the server directory, but GoogleAuthenticator shouldn't necessarily live there.
We also need to get a URL from GoogleAuthenticator to show on the login page.
We hack around this, I'm not super happy with it, but it works for the moment.
Open to better ideas about how this should work.

Move the YAML config out of the logrole_server binary and into the config
directory. Add Policy to it, and a custom parser for UserSettings. These let
other Go code load a Logrole YAML file, if they want.

Add more documentation around possibly-confusing settings. Document how to get
a Google client ID and client secret.

Fix errors in Google authentication, and add a whole bunch of tests around
policies, Google auth, and Basic auth. Removes some unused code that set
a global map in the config directory.

Add tools in write_config_from_env to download a policy file from a URL (for
Heroku deployment, if you can't include the permissions as part of the Git
repo).

## 0.56

Highlight Call list rows in red if the call ended unsuccessfully.

## 0.55

Show error/warning information about a Call on the instance page.

Messages that resulted in an error are highlighted in red.

## 0.54

You can configure timezones via config.yml, and the timezones in the menu bar
are now dynamic.

## 0.51

Gzip static files so they get sent to the client more quickly. It would be nice
to also gzip the HTML, but this would be vulnerable to BEAST/CRIME attacks on
SSL.

## 0.50

Add a Conference instance view.

## 0.48

Renamed the binaries from `server` and `write_config_from_env` to
`logrole_server` and `logrole_write_config_from_env` to avoid conflicts with
other Go binaries.

Add `google_allowed_domains` config variable to restrict access to email
addresses that are part of a certain domain.


config/settings.go
package config

import (
	"encoding/hex"
	"errors"
	"fmt"
	"io/ioutil"
	"net"
	"net/mail"
	"time"

	log "github.com/inconshreveable/log15"
	"github.com/kevinburke/handlers"
	twilio "github.com/kevinburke/twilio-go"
	"github.com/saintpete/logrole/services"
	yaml "gopkg.in/yaml.v2"
)

const DefaultPort = "4114"
const DefaultPageSize = 50

// DefaultTimezones are a user's options if no timezones are configured. These
// correspond to the 4 timezones in the USA, west to east.
var DefaultTimezones = []string{
	"America/Los_Angeles",
	"America/Denver",
	"America/Chicago",
	"America/New_York",
}

// DefaultMaxResourceAge allows all resources to be fetched. The company was
// founded in 2008, so there should definitely be no resources created in the
// 1980's.
var DefaultMaxResourceAge = time.Since(time.Date(1990, 1, 1, 0, 0, 0, 0, time.UTC))

var missingGoogleCredentials = errors.New("Cannot use google auth without a Client ID and Client Secret. To configure a Client ID and Secret, see https://github.com/saintpete/logrole/blob/master/docs/google.md.")

// FileConfig defines the settings you can load from a YAML configuration file.
// Load configuration from a YAML file into a FileConfig struct, then call
// NewSettingsFromConfig to get a Settings object.
//
// All of the types and values here should be representable in a YAML file.
type FileConfig struct {
	Port       string `yaml:"port"`
	AccountSid string `yaml:"twilio_account_sid"`
	AuthToken  string `yaml:"twilio_auth_token"`

	Realm services.Rlm `yaml:"realm"`
	// Default timezone for dates/times in the UI
	Timezone string `yaml:"default_timezone"`
	// List of timezones a user can choose in the UI
	Timezones  []string `yaml:"timezones"`
	PublicHost string   `yaml:"public_host"`

	// IP subnets that are allowed to visit the site. THIS IS NOT A SECURITY
	// FEATURE. IP ADDRESSES ARE EASILY SPOOFED, AND YOUR IP ADDRESS IS EASILY
	// DISCOVERABLE. To determine a user's ip address, we check the first value
	// in a X-Forwarded-For header, or the RemoteHost value of a http.Request.
	//
	// If you have an IPv4 address, the subnet for *only* that address is
	// "A.B.C.D/32". The recommended smallest subnet for IPv6 is /64.
	IPSubnets []string `yaml:"ip_subnets"`

	PageSize       uint          `yaml:"page_size"`
	SecretKey      string        `yaml:"secret_key"`
	MaxResourceAge time.Duration `yaml:"max_resource_age"`

	// Need a pointer to a boolean here since we want to be able to distinguish
	// "false" from "omitted"
	ShowMediaByDefault *bool `yaml:"show_media_by_default,omitempty"`

	EmailAddress string `yaml:"email_address"`

	ErrorReporter      string `yaml:"error_reporter,omitempty"`
	ErrorReporterToken string `yaml:"error_reporter_token,omitempty"`

	AuthScheme string `yaml:"auth_scheme"`
	User       string `yaml:"basic_auth_user"`
	Password   string `yaml:"basic_auth_password"`

	GoogleClientID       string   `yaml:"google_client_id"`
	GoogleClientSecret   string   `yaml:"google_client_secret"`
	GoogleAllowedDomains []string `yaml:"google_allowed_domains"`

	PolicyFile string `yaml:"policy_file"`
	Policy     *Policy

	Debug bool `yaml:"debug"`
}

// Settings are used to configure a Server and apply to all of the website's
// users.
type Settings struct {
	Logger log.Logger

	// The host the user visits to get to this site.
	PublicHost string

	// Whether to allow HTTP traffic.
	AllowUnencryptedTraffic bool
	Client                  *twilio.Client

	// LocationFinder determines the correct timezone to display for a given
	// request, based on the default and a user's TZ cookie (if present).
	LocationFinder services.LocationFinder

	// How many messages to display per page.
	PageSize uint

	// Used to encrypt next page URI's and sessions. See
	// https://github.com/saintpete/logrole/blob/master/docs/settings.md#secret-key
	SecretKey *[32]byte

	// Don't show resources that are older than this age. Set to a very high
	// value to show all resources.
	MaxResourceAge time.Duration

	// Should a user have to click a button to view media attached to a MMS?
	ShowMediaByDefault bool

	// Email address for server errors / "contact me" on error pages.
	Mailto *mail.Address

	// Error reporter. This must not be nil; set to NoopErrorReporter to ignore
	// errors.
	Reporter services.ErrorReporter

	// The authentication scheme.
	Authenticator Authenticator

	// THIS IS NOT A SECURITY FEATURE AND SHOULD NOT BE RELIED ON FOR IP
	// WHITELISTING.
	IPSubnets []*net.IPNet
}

var errWrongLength = errors.New("Secret key has wrong length. Should be a 64-byte hex string")

// getSecretKey produces a valid [32]byte secret key or returns an error. If
// hexKey is the empty string, a valid 32 byte key will be randomly generated
// and returned. If hexKey is invalid, an error is returned.
func getSecretKey(hexKey string) (*[32]byte, error) {
	if hexKey == "" {
		return services.NewRandomKey(), nil
	}

	if len(hexKey) != 64 {
		return nil, errWrongLength
	}
	secretKeyBytes, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, err
	}
	secretKey := new([32]byte)
	copy(secretKey[:], secretKeyBytes)
	return secretKey, nil
}

// NewSettingsFromConfig creates a new Settings object from the given
// FileConfig, or an error.
//
// Pass a log.Logger to configure how messages are logged. If the Logger is
// nil, github.com/kevinburke/handlers.Logger will be used.
func NewSettingsFromConfig(c *FileConfig, l log.Logger) (settings *Settings, err error) {
	defer func() {
		if r := recover(); r != nil {
			if c.Debug {
				panic(r)
			}
			err = fmt.Errorf("Panic in NewSettings: %#v", r)
			settings = nil
			return
		}
	}()
	if l == nil {
		if c.Debug {
			l = handlers.NewLoggerLevel(log.LvlDebug)
		} else {
			l = handlers.Logger
		}
	}
	if c.Policy != nil && c.PolicyFile != "" {
		return nil, errors.New("Cannot define both policy and a policy_file")
	}
	allowHTTP := false
	if c.Realm == services.Local {
		allowHTTP = true
	}
	if c.SecretKey == "" {
		l.Warn("No secret key provided, generating random secret key. Sessions won't persist across restarts")
	}
	secretKey, err := getSecretKey(c.SecretKey)
	if err != nil {
		return nil, err
	}
	if c.MaxResourceAge == 0 {
		c.MaxResourceAge = DefaultMaxResourceAge
	}
	var address *mail.Address
	if c.EmailAddress != "" {
		address, err = mail.ParseAddress(c.EmailAddress)
		if err != nil {
			return nil, fmt.Errorf("Couldn't parse email address: %v", err)
		}
	}
	if c.ErrorReporter != "" {
		if !services.IsRegistered(c.ErrorReporter) {
			l.Warn("Unknown error reporter, using the noop reporter", "name", c.ErrorReporter)
		}
	}
	reporter := services.GetReporter(c.ErrorReporter, c.ErrorReporterToken)

	if c.PolicyFile != "" {
		// we checked above that Policy is nil in this case
		data, err := ioutil.ReadFile(c.PolicyFile)
		if err != nil {
			l.Error("Couldn't load permission file", "loc", c.PolicyFile)
			return nil, err
		}
		var policy Policy
		if err := yaml.Unmarshal(data, &policy); err != nil {
			l.Error("Couldn't parse policy file", "err", err, "loc", c.PolicyFile)
			return nil, err
		}
		c.Policy = &policy
	}

	if c.Policy != nil {
		if err := validatePolicy(c.Policy); err != nil {
			l.Error("Couldn't validate policy", "err", err)
			return nil, err
		}
	}
	var authenticator Authenticator
	switch c.AuthScheme {
	case "":
		l.Warn("Disabling basic authentication")
		authenticator = &NoopAuthenticator{User: DefaultUser}
	case "basic":
		if c.User == "" || c.Password == "" {
			return nil, errors.New("Cannot use basic auth without a username or password, set a basic_auth_user")
		}
		ba := NewBasicAuthAuthenticator("logrole")
		ba.AddUserPassword(c.User, c.Password)
		authenticator = ba
	case "google":
		if c.GoogleClientID == "" || c.GoogleClientSecret == "" {
			return nil, missingGoogleCredentials
		}
		var baseURL string
		if allowHTTP {
			baseURL = "http://" + c.PublicHost
		} else {
			baseURL = "https://" + c.PublicHost
		}
		gauthenticator := NewGoogleAuthenticator(l, c.GoogleClientID, c.GoogleClientSecret, baseURL, c.GoogleAllowedDomains, secretKey)
		gauthenticator.AllowUnencryptedTraffic = allowHTTP
		authenticator = gauthenticator
	default:
		return nil, fmt.Errorf("Unknown auth scheme: %s", c.AuthScheme)
	}
	authenticator.SetPolicy(c.Policy)
	client := twilio.NewClient(c.AccountSid, c.AuthToken, nil)
	if c.Timezone == "" {
		l.Info("No timezone provided, defaulting to UTC")
	}
	locationFinder, err := services.NewLocationFinder(c.Timezone)
	if err != nil {
		return nil, fmt.Errorf("Couldn't find timezone %s: %s", c.Timezone, err.Error())
	}
	tzs := DefaultTimezones
	if len(c.Timezones) > 0 {
		tzs = c.Timezones
	}
	for _, timezone := range tzs {
		if ok := locationFinder.AddLocation(timezone); !ok {
			l.Warn("Couldn't add location", "tz", timezone)
		}
	}
	var nets []*net.IPNet
	if c.IPSubnets == nil {
		nets = make([]*net.IPNet, 0)
	} else {
		nets = make([]*net.IPNet, len(c.IPSubnets))
		for i, ipStr := range c.IPSubnets {
			_, n, err := net.ParseCIDR(ipStr)
			if err != nil {
				l.Error("Couldn't parse IP subnet", "err", err, "ip", ipStr)
				return nil, err
			}
			nets[i] = n
		}
	}

	// TODO
	if c.PageSize == 0 {
		c.PageSize = DefaultPageSize
	}
	if c.PageSize > 1000 {
		return nil, fmt.Errorf("Maximum allowable page size is 1000, got %d", c.PageSize)
	}
	if c.ShowMediaByDefault == nil {
		b := true
		c.ShowMediaByDefault = &b
	}

	settings = &Settings{
		Logger:                  l,
		AllowUnencryptedTraffic: allowHTTP,
		Client:                  client,
		LocationFinder:          locationFinder,
		PublicHost:              c.PublicHost,
		PageSize:                c.PageSize,
		SecretKey:               secretKey,
		MaxResourceAge:          c.MaxResourceAge,
		ShowMediaByDefault:      *c.ShowMediaByDefault,
		Mailto:                  address,
		Reporter:                reporter,
		Authenticator:           authenticator,
		IPSubnets:               nets,
	}
	return
}
